### ISSUE TITLE
First request takes as long as the connectTimeout

### ISSUE THREAD

**Issue by houmie** [square/okhttp#6486]
The first request to any backend takes as long as the connectTimeout is defined (5 seconds). Any subsequent calls go then really fast. Why is the first call waiting until the timeout exceeds before it succeeds?

I may have found a bug. This is independent from API. Because when I change .connectTimeout(5, TimeUnit.SECONDS) to 10 seconds. Then the first call takes 10 seconds. And if I lower it to 1 second, then it takes 1 second to finish the first call.

I even tried to disable proxy .proxy(Proxy.NO_PROXY) but it didn't help.

```
import okhttp3.OkHttpClient
import retrofit2.Call
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import java.util.concurrent.TimeUnit

object ApiService {
    private val SERVICE : ApiInterface

    init {
        val okHttpClient = OkHttpClient().newBuilder()
            .connectTimeout(5, TimeUnit.SECONDS)
            .readTimeout(1, TimeUnit.SECONDS)
            .writeTimeout(1, TimeUnit.SECONDS)
            .build()
        val retrofit = Retrofit.Builder()
            .baseUrl(Config.BASE_URL)
            .client(okHttpClient)
            .addConverterFactory(MoshiConverterFactory.create())
            .build()
        SERVICE = retrofit.create(ApiInterface::class.java)
    }

    fun getServerStatus() : Call<Map<String, ServerJson>> = SERVICE.getServers()

}
```

I think this goes back to proxy and dns resolving issue over https.

So I used this to override the system dns.
```
implementation('com.squareup.okhttp3:okhttp-dnsoverhttps:4.9.0')
```

AND also disabled proxy.

```
init {
        val appCache = Cache(File("cacheDir", "okhttpcache"), 10 * 1024 * 1024)
        val bootstrapClient = OkHttpClient.Builder()
            .cache(appCache)
            .proxy(Proxy.NO_PROXY)
            .build()
        val dns = DnsOverHttps.Builder()
            .client(bootstrapClient)
            .url("https://dns.google/dns-query".toHttpUrl())
            .bootstrapDnsHosts(InetAddress.getByName("8.8.4.4"), InetAddress.getByName("8.8.8.8"))
            .build()
        val okHttpClient = OkHttpClient().newBuilder()
            .connectTimeout(60, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(60, TimeUnit.SECONDS)
            .dns(dns)
            .proxy(Proxy.NO_PROXY)
            .build()
        val retrofit = Retrofit.Builder()
            .baseUrl(Config.BASE_URL)
            .client(okHttpClient)
            .addConverterFactory(MoshiConverterFactory.create())
            .build()
        SERVICE = retrofit.create(ApiInterface::class.java)
    }
```

This actually works as expected (no waiting time when doing the first call) when I build the app and run it on the phone.
However when the app goes to background and comes back, it loses it.
The next call takes again as long as the connectTimeout.

I'm using the latest version:
```
implementation('com.squareup.okhttp3:okhttp')
```

**Comment by swankjesse** [comment]
Does the call it succeed or fail after the timeout? Typically timeouts trigger failures, but perhaps something is retrying?

What's your proxy server? It might be that the first call ends up putting the proxy on the list of routes to avoid.

**Comment by houmie** [comment]
It succeeds. I don't have any proxy server set up.  I call a plain Restful API.
I did further research and it seems more people have experienced this issue in the past: https://github.com/square/retrofit/issues/1489#issuecomment-698785683

If you scroll up some claim setting a custom DNS has resolved it. But I'm already doing that. Very strange.

**Comment by yschimke** [comment]
@houmie could you provide output for https://square.github.io/okhttp/4.x/okhttp-logging-interceptor/okhttp3.logging/-logging-event-listener/

https://square.github.io/okhttp/events/

I'd like to understand whether multiple routes are involved or where the delay is?

**Comment by yschimke** [comment]
One example scenario would be if you have a MITM as the first DNS result, which timeouts, then the route is avoided on all further calls.

**Comment by houmie** [comment]
@yschimke Certainly, thank you for your help.
I believe, it is trying to connect first via IPv6, then it fails and switches to IPv4. 
The remote API is closed to IPv6 traffic.  My home internet provider provides both IPv6 and IPv4. Maybe this is why.
What do you think?

I have added it like this:
```
val listener = LoggingEventListener.Factory()
val okHttpClient = OkHttpClient().newBuilder()
            .eventListenerFactory(listener)
            .connectTimeout(60, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .writeTimeout(60, TimeUnit.SECONDS)
```

The outcome is very interesting:

```
I/okhttp.OkHttpClient: [60161 ms] connectFailed: null java.net.SocketTimeoutException: failed to connect to /2a05:d012:1a7:1111:d9e4:aaaa:18e1:6cdc (port 443) from /2a01:4b00:1111:6d00:aaaa:71ff:fe62:d04c (port 47504) after 59999ms
I/okhttp.OkHttpClient: [60163 ms] connectStart: /18.133.100.100:443 DIRECT
I/okhttp.OkHttpClient: [60185 ms] secureConnectStart
I/okhttp.OkHttpClient: [60258 ms] secureConnectEnd: Handshake{tlsVersion=TLS_1_2 cipherSuite=TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 peerCertificates=[CN=api-1.mydomain.net, CN=R3, O=Let's Encrypt, C=US, CN=DST Root CA X3, O=Digital Signature Trust Co.] localCertificates=[]}
I/okhttp.OkHttpClient: [60258 ms] connectEnd: http/1.1
    [60259 ms] connectionAcquired: Connection{api-1.mydomain.net:443, proxy=DIRECT hostAddress=/18.133.100.100:443 cipherSuite=TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 protocol=http/1.1}
I/okhttp.OkHttpClient: [60262 ms] requestHeadersStart
I/okhttp.OkHttpClient: [60262 ms] requestHeadersEnd
    [60263 ms] requestBodyStart
I/okhttp.OkHttpClient: [60264 ms] requestBodyEnd: byteCount=26
I/okhttp.OkHttpClient: [60283 ms] responseHeadersStart
    [60284 ms] responseHeadersEnd: Response{protocol=http/1.1, code=400, message=BAD REQUEST,  #url=https://api-1.mydomain.net/v7/signin}
I/okhttp.OkHttpClient: [60285 ms] responseBodyStart
    [60285 ms] responseBodyEnd: byteCount=71
I/okhttp.OkHttpClient: [60285 ms] connectionReleased
    [60285 ms] callEnd
```

I have anonymised the domain , IPv4 and IPv6 above.
Thanks

**Comment by yschimke** [comment]
Ideally the fix would be in OkHttp - Happy Eyeballs https://github.com/square/okhttp/issues/506. But that isn't happening soon.

As a high touch solution you could disable IPv6 either on your device, or https://www.cactusvpn.com/tutorials/how-to-disable-ipv6-on-android/ or look into why your network dns or the remote service is advertising IPv6.  It's generally a network misconfiguration somewhere if you are seeing IPv6 addresses from DNS that are getting blackholed.

The pragmatic solution is to implement a Dns decorator to filter out IPv6, but it's a bandaid https://github.com/yschimke/okurl/blob/e307022667e2beb474309af5b350cd241f2a9045/src/main/kotlin/com/baulsupp/okurl/network/DnsSelector.kt

But honestly it's good to know that OkHttp is doing the correct thing here and repairing the connection.

**Comment by yschimke** [comment]
narrator: "It was dns."

**Comment by houmie** [comment]
@yschimke Thanks Yuri, the DNS for IPv6 is now removed from the API and it works without any delay!!!  

Thanks for suggesting ```logging-event-listener```, it's a very useful tool for troubleshooting.

**Comment by willqz** [comment]
> The first request to any backend takes as long as the connectTimeout is defined (5 seconds). Any subsequent calls go then really fast. Why is the first call waiting until the timeout exceeds before it succeeds?
> 
> I may have found a bug. This is independent from API. Because when I change .connectTimeout(5, TimeUnit.SECONDS) to 10 seconds. Then the first call takes 10 seconds. And if I lower it to 1 second, then it takes 1 second to finish the first call.
> 
> I even tried to disable proxy .proxy(Proxy.NO_PROXY) but it didn't help.
> 
> ```
> import okhttp3.OkHttpClient
> import retrofit2.Call
> import retrofit2.Retrofit
> import retrofit2.converter.moshi.MoshiConverterFactory
> import java.util.concurrent.TimeUnit
> 
> object ApiService {
>     private val SERVICE : ApiInterface
> 
>     init {
>         val okHttpClient = OkHttpClient().newBuilder()
>             .connectTimeout(5, TimeUnit.SECONDS)
>             .readTimeout(1, TimeUnit.SECONDS)
>             .writeTimeout(1, TimeUnit.SECONDS)
>             .build()
>         val retrofit = Retrofit.Builder()
>             .baseUrl(Config.BASE_URL)
>             .client(okHttpClient)
>             .addConverterFactory(MoshiConverterFactory.create())
>             .build()
>         SERVICE = retrofit.create(ApiInterface::class.java)
>     }
> 
>     fun getServerStatus() : Call<Map<String, ServerJson>> = SERVICE.getServers()
> 
> }
> ```
> 
> I think this goes back to proxy and dns resolving issue over https.
> 
> So I used this to override the system dns.
> 
> ```
> implementation('com.squareup.okhttp3:okhttp-dnsoverhttps:4.9.0')
> ```
> 
> AND also disabled proxy.
> 
> ```
> init {
>         val appCache = Cache(File("cacheDir", "okhttpcache"), 10 * 1024 * 1024)
>         val bootstrapClient = OkHttpClient.Builder()
>             .cache(appCache)
>             .proxy(Proxy.NO_PROXY)
>             .build()
>         val dns = DnsOverHttps.Builder()
>             .client(bootstrapClient)
>             .url("https://dns.google/dns-query".toHttpUrl())
>             .bootstrapDnsHosts(InetAddress.getByName("8.8.4.4"), InetAddress.getByName("8.8.8.8"))
>             .build()
>         val okHttpClient = OkHttpClient().newBuilder()
>             .connectTimeout(60, TimeUnit.SECONDS)
>             .readTimeout(60, TimeUnit.SECONDS)
>             .writeTimeout(60, TimeUnit.SECONDS)
>             .dns(dns)
>             .proxy(Proxy.NO_PROXY)
>             .build()
>         val retrofit = Retrofit.Builder()
>             .baseUrl(Config.BASE_URL)
>             .client(okHttpClient)
>             .addConverterFactory(MoshiConverterFactory.create())
>             .build()
>         SERVICE = retrofit.create(ApiInterface::class.java)
>     }
> ```
> 
> This actually works as expected (no waiting time when doing the first call) when I build the app and run it on the phone. However when the app goes to background and comes back, it loses it. The next call takes again as long as the connectTimeout.
> 
> I'm using the latest version:
> 
> ```
> implementation('com.squareup.okhttp3:okhttp')
> ```

Now working perfect!!!

**Comment by jahertor** [comment]
The DnsOverHttps didn't worked to me.

The problem is with mixed IPv4+IPv6 networks. It's solved on Okhttp 5.0.0-alpha (https://square.github.io/okhttp/changelogs/changelog/#version-500-alpha4)

I got the key here: https://github.com/facebook/react-native/issues/28283#issuecomment-1416082231

```
dependencies {
    // others dependencies
  
    implementation 'com.squareup.okhttp3:logging-interceptor:5.0.0-alpha.11'
    implementation 'com.squareup.okhttp3:okhttp:5.0.0-alpha.11'
    implementation 'com.squareup.okhttp3:okhttp-urlconnection:5.0.0-alpha.11'
}
```

**Comment by LunevNF** [comment]
Problem repeats even after update to v5.0.+ :(
I already try to add dns, but problem not fixed
